// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: testdata.proto

package protobuf

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on SampleRecord with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SampleRecord) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SampleRecord with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SampleRecordMultiError, or
// nil if none found.
func (m *SampleRecord) ValidateAll() error {
	return m.validate(true)
}

func (m *SampleRecord) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	// no validation rules for Email

	// no validation rules for DisplayName

	// no validation rules for Password

	// no validation rules for Age

	// no validation rules for Weight

	// no validation rules for EnableSingleSignOn

	if len(errors) > 0 {
		return SampleRecordMultiError(errors)
	}

	return nil
}

// SampleRecordMultiError is an error wrapping multiple validation errors
// returned by SampleRecord.ValidateAll() if the designated constraints aren't met.
type SampleRecordMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SampleRecordMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SampleRecordMultiError) AllErrors() []error { return m }

// SampleRecordValidationError is the validation error returned by
// SampleRecord.Validate if the designated constraints aren't met.
type SampleRecordValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SampleRecordValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SampleRecordValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SampleRecordValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SampleRecordValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SampleRecordValidationError) ErrorName() string { return "SampleRecordValidationError" }

// Error satisfies the builtin error interface
func (e SampleRecordValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSampleRecord.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SampleRecordValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SampleRecordValidationError{}

// Validate checks the field values on SampleRecordWithError with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SampleRecordWithError) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SampleRecordWithError with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SampleRecordWithErrorMultiError, or nil if none found.
func (m *SampleRecordWithError) ValidateAll() error {
	return m.validate(true)
}

func (m *SampleRecordWithError) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetBase()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SampleRecordWithErrorValidationError{
					field:  "Base",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SampleRecordWithErrorValidationError{
					field:  "Base",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetBase()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SampleRecordWithErrorValidationError{
				field:  "Base",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for ErrorContent

	if len(errors) > 0 {
		return SampleRecordWithErrorMultiError(errors)
	}

	return nil
}

// SampleRecordWithErrorMultiError is an error wrapping multiple validation
// errors returned by SampleRecordWithError.ValidateAll() if the designated
// constraints aren't met.
type SampleRecordWithErrorMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SampleRecordWithErrorMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SampleRecordWithErrorMultiError) AllErrors() []error { return m }

// SampleRecordWithErrorValidationError is the validation error returned by
// SampleRecordWithError.Validate if the designated constraints aren't met.
type SampleRecordWithErrorValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SampleRecordWithErrorValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SampleRecordWithErrorValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SampleRecordWithErrorValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SampleRecordWithErrorValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SampleRecordWithErrorValidationError) ErrorName() string {
	return "SampleRecordWithErrorValidationError"
}

// Error satisfies the builtin error interface
func (e SampleRecordWithErrorValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSampleRecordWithError.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SampleRecordWithErrorValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SampleRecordWithErrorValidationError{}
